mautic/mautic
# First segment filter glue fixed

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [✅]
| New feature/enhancement? (use the a.x branch)      | [❌]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [❌]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      |  
| Related developer documentation PR URL |  
| Issue(s) addressed                     |  



#### Description:



This PR fixes a bug that causes an invalid rebuild logic applied for segments containing multiple "Segment Membership" filters placed at the top of the filter list. See the recordings:

https://user-images.githubusercontent.com/6388925/231696652-9f03fbb3-df94-421d-a90b-bda70dff327a.mov

---
The rebuild query for the above segment should look like:
```sql
SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId
FROM (SELECT DISTINCT l.id as leadIdPrimary
    FROM leads l
    WHERE
            (NOT EXISTS(SELECT null FROM leads par1 LEFT JOIN lead_lists_leads par4 ON par1.id = par4.lead_id and par4.leadlist_id = 4 and (par4.manually_removed = 1)
            WHERE (l.id = par1.id) AND (((par1.country = 'Japan'))) /* ... */ ))
        AND
            (NOT EXISTS(SELECT null FROM leads par6 LEFT JOIN lead_lists_leads par9 ON par6.id = par9.lead_id and par9.leadlist_id = 3 and (par9.manually_removed = 1)
            WHERE (l.id = par6.id) AND (((par6.country = 'United States'))) /* ... */ ))
        AND
            (EXISTS(SELECT null FROM leads parc LEFT JOIN lead_lists_leads parf ON parc.id = parf.lead_id and parf.leadlist_id = 1 and (parf.manually_removed = 1)
            WHERE (l.id = parc.id) AND (((parc.country = 'Canada'))) /* ... */ ))
        AND (l.id NOT IN (SELECT parh.lead_id FROM lead_lists_leads parh WHERE parh.leadlist_id = '2'))) sss;
```

But it is generated as follows:
```sql
SELECT count(leadIdPrimary) count, max(leadIdPrimary) maxId, min(leadIdPrimary) minId
FROM (SELECT DISTINCT l.id as leadIdPrimary
    FROM leads l
    WHERE (
        (NOT EXISTS(SELECT null FROM leads par1 LEFT JOIN lead_lists_leads par4 ON par1.id = par4.lead_id and par4.leadlist_id = 4 and (par4.manually_removed = 1)
             WHERE (l.id = par1.id) AND (((par1.country = 'Japan'))) /* ... */ ))
        OR
        (
             (NOT EXISTS(SELECT null FROM leads par6 LEFT JOIN lead_lists_leads par9 ON par6.id = par9.lead_id and par9.leadlist_id = 3 and (par9.manually_removed = 1)
                 WHERE (l.id = par6.id) AND (((par6.country = 'United States'))) /* ... */ ))
             AND
            (EXISTS(SELECT null FROM leads parc LEFT JOIN lead_lists_leads parf ON parc.id = parf.lead_id and parf.leadlist_id = 1 and (parf.manually_removed = 1)
                WHERE (l.id = parc.id) AND (((parc.country = 'Canada'))) /* ... */ ))
        )
    )
    AND (l.id NOT IN (SELECT parh.lead_id FROM lead_lists_leads parh WHERE parh.leadlist_id = '2'))) sss;
```

#### Steps to test this PR:


1. Create a segment with the following filters:
    - Segment membership -> Excluding -> Choose 2 segments
    - **Or** -> Segment membership -> Including -> Choose 1 segments (make sure the "or" glue is selected)
2. Drag the first segment filter below the second one.
3. You should see the two filters connected with an "and" glue.
4. Save the segment.
5. Make sure the segment contains appropriate contacts after a full rebuilt is done.
